### PR TITLE
ci(msys): faster builds using msys2-setup

### PR DESCRIPTION
**Pull request by kylo252**
It's currently ~15min just to configure and prepare the msys2 environment https://github.com/neovim/neovim/runs/6269540875?check_suite_focus=true#step:4:251
whereas nearly all the other jobs are completely done under 10min

This is now using a github-action https://github.com/msys2/setup-msys2 which is under the official msys2 org, and they claim to have solved those update issues, so I thought I could give it a try.

Related #18384 ~~15490~~

### CODE REVIEW

**Comment by zeertzjq** [comment]
Did you link the wrong issue number?

File: ci/build.ps1 [hunk 1]
```diff
@@ -64,19 +64,8 @@ if ($compiler -eq 'MINGW') {
...
-  # Avoid pacman "warning" which causes non-zero return code. https://github.com/open62541/open62541/issues/2068
-  & C:\msys64\usr\bin\mkdir -p /var/cache/pacman/pkg
-
-  # Build third-party dependencies
-  C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm -Syu" ; exitIfFailed
-  # Update again in case updating pacman changes pacman.conf
-  C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm -Syu" ; exitIfFailed
-  C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm --needed -S $mingwPackages" ; exitIfFailed
 }
+
```
Comment (justinmk) [inline ci/build.ps1:68]: delete

**Comment by kylo252** [comment]
> Fork the repo and test the PR on your own fork.

It's already on my fork, but I keep getting different kinds of errors, probably because it's on a fork and/or missing github tokens, see https://github.com/kylo252/neovim/actions/runs/2264393494

 I guess it's fastest just to mark it as ready instead.

**Comment by dundargoc** [comment]
> It's already on my fork, but I keep getting different kinds of errors, probably because it's on a fork and/or missing github tokens, see [kylo252/neovim/actions/runs/2264393494](https://github.com/kylo252/neovim/actions/runs/2264393494)

You need to add quotes around MINGW_64-gcov, that's why it's complaining.

**Comment by kylo252** [comment]
> You need to add quotes around MINGW_64-gcov, that's why it's complaining.

thanks! testing to see if it works. 

the good news is that it already seems promising, because the job got to the cmake part in ~3min 😃

File: ci/build.ps1 [hunk 2]
```diff
@@ -61,21 +61,14 @@ if ($compiler -eq 'MINGW') {
   # in MSYS2, but we cannot build inside the MSYS2 shell.
   $cmakeGenerator = 'Ninja'
   $cmakeGeneratorArgs = '-v'
-  $mingwPackages = @('ninja', 'cmake', 'diffutils').ForEach({
-    "mingw-w64-$arch-$_"
-  })
 
+  if (Get-Command -Name "msys2" -ErrorAction SilentlyContinue) {
+    $MINGW_PREFIX = (& msys2 -c 'cygpath -w /') + "mingw$bits\bin"
```
Comment (jamessan) [inline ci/build.ps1:66]: Should this be

```suggestion
    $MINGW_PREFIX = (& msys2 -c 'cygpath -w /') + "mingw$bits/bin"
```

Comment (kylo252) [reply ci/build.ps1:66]: this should be fine
```console
$ /usr/bin/cygpath -w /
C:\msys64\
```

Comment (kylo252) [reply ci/build.ps1:66]: the current problem is currently here:
https://github.com/neovim/neovim/runs/6291974592?check_suite_focus=true#step:5:74
which for some reason is unable to find `ninja`, but checking previous builds seems to indicate that this step is actually using `make`
https://github.com/neovim/neovim/runs/6308146076?check_suite_focus=true#step:4:189

this is probably related to the shared cache?

Comment (kylo252) [reply ci/build.ps1:73]: so far it seems that ruby might be the culprit here, but moving that block further down should fix the issue

**Comment by kylo252** [comment]
@justinmk, this PR is basically complete pending the decision on #18494

**Comment by justinmk** [comment]
I guess it's taking longer beacause of the hang? If it's only 1 or 2 tests we can skip them on msys for (thanks for creating a tracking issue).

Then let's see if this is actually faster. On current master, mingw build takes 33 minutes.

Would also expect to see some deleted code here... currently it just looks like more code.

**Comment by justinmk** [comment]
In https://github.com/neovim/neovim/issues/18551 we are considering whether we should eliminate the mingw/msys2 CI job.

**Comment by justinmk** [comment]
Conclusion from https://github.com/neovim/neovim/issues/18551 is that we will remove msys2 / mingw CI.

**Pull request closed without merging**